tests/realtikvtest/addindextest3/ingest_test.go

```diff
@@ -776,3 +776,60 @@ func TestIssue55808(t *testing.T) {
 	err := tk.ExecToErr("alter table t add index idx(a);")
 	require.ErrorContains(t, err, "injected error")
 }
+
+func TestAddIndexBackfillLostTempIndexValues(t *testing.T) {
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("drop database if exists addindexlit;")
+	tk.MustExec("create database addindexlit;")
+	tk.MustExec("use addindexlit;")
+	tk.MustExec(`set global tidb_ddl_enable_fast_reorg=on;`)
+	tk.MustExec("set global tidb_enable_dist_task = 0;")
+
+	tk.MustExec("create table t(id int primary key, b int not null default 0);")
+	tk.MustExec("insert into t values (1, 0);")
+
+	tk1 := testkit.NewTestKit(t, store)
+	tk1.MustExec("use addindexlit;")
+
+	ddl.MockDMLExecutionBeforeScan = func() {
+		_, err := tk1.Exec("insert into t values (2, 0);")
+		assert.NoError(t, err)
+		_, err = tk1.Exec("delete from t where id = 1;")
+		assert.NoError(t, err)
+		_, err = tk1.Exec("insert into t values (3, 0);")
+		assert.NoError(t, err)
+		_, err = tk1.Exec("delete from t where id = 2;")
+		assert.NoError(t, err)
+	}
+
+	ingest.MockDMLExecutionStateBeforeImport = func() {
+		_, err := tk1.Exec("insert into t(id) values (4);")
+		assert.NoError(t, err)
+		_, err = tk1.Exec("delete from t where id = 3;")
+		assert.NoError(t, err)
+	}
+	var rows [][]any
+	ddl.MockDMLExecutionStateBeforeMerge = func() {
+		rows = tk1.MustQuery("select * from t use index();").Rows()
+		_, err := tk1.Exec("insert into t values (3, 0);")
+		assert.NoError(t, err)
+	}
+
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/pkg/ddl/skipReorgWorkForTempIndex", "return(false)"))
```

Comment:
You might be able to use `testfailpoint` pkg instead of the error checks + cleanup for `failpoint` pkg.